usr/share/rear/lib/global-functions.sh

```diff
@@ -369,6 +369,10 @@ function scheme_accepts_files() {
 ### only that it can be mounted (use mount_url() first)
 function scheme_supports_filesystem() {
     local scheme=$1
+    # Return false if scheme is empty or blank (e.g. when OUTPUT_URL is unset or empty or blank)
+    # cf. https://github.com/rear/rear/issues/2676
+    # and https://github.com/rear/rear/issues/2667#issuecomment-914447326
+    test $scheme || return 1
```

Comment:
don't you also need to quote `$scheme` in callers? In principle it should work even without quoting, because this function takes only one argument, but it will cause problems when `set -eu` is used (`$1` can be undefined).

Reply:
for consistency, I think that the same thing should be added to `scheme_accepts_files()` above.

Reply:
And it should also fail if scheme is more than one word
```
# test $scheme && echo y || echo n
-bash: test: foo: unary operator expected
n
```
That doesn't look nice but actually it is perfectly right
to have that bash stderr message.

Reply:
Cf.  "Beware of the emptiness" in
https://github.com/rear/rear/wiki/Coding-Style

Reply:
> I use it without quotes to not let it falsely succeed if $scheme is blank

Here it is fine. I meant, one should use quotes in the callers of `scheme_supports_filesystem`.

Reply:
Ah! - now I understand what you meant:
```
# function foo() { echo $1 ; echo $2 ; }

# ( set -eu ; foo 1 2  && echo y || echo n ; echo hello )
1
2
y
hello

# ( set -eu ; foo 1 2 3 && echo y || echo n ; echo hello )
1
2
y
hello

# ( set -eu ; foo 1 && echo y || echo n ; echo hello )
1
bash: $2: unbound variable

# ( set -eu ; foo && echo y || echo n ; echo hello )
bash: $1: unbound variable
```
So when `set -eu`
calling functions with too many arguments works but
calling functions with too few arguments exits bash.

Interestingly bash parameter expansion with assigning default values
```
# function foo() { echo ${1:=a} ; echo ${2:=b} ; }

# ( set -eu ; foo 1 && echo y || echo n ; echo hello )
1
bash: $2: cannot assign in this way

# function foo() { echo ${1:=} ; echo ${2:=} ; }

# ( set -eu ; foo 1 && echo y || echo n ; echo hello )
1
bash: $2: cannot assign in this way
```
does not help that calling functions with too few arguments exits bash
but bash parameter expansion with only using default values
helps against exiting bash when calling functions with too few arguments
```
# function foo() { echo ${1:-a} ; echo ${2:-b} ; }

# ( set -eu ; foo 1 && echo y || echo n ; echo hello )
1
b
y
hello

# function foo() { echo ${1:-} ; echo ${2:-} ; }

# ( set -eu ; foo 1 && echo y || echo n ; echo hello )
1

y
hello
```
so I will implement the latter method
for scheme_supports_filesystem() and scheme_accepts_files()
because in general it is better to improve the functions instead of all its callers.

Reply:
Implemented via
https://github.com/rear/rear/pull/2675/commits/2965acf0a3589722d5be72b2734b03baf4084fe8

Reply:
I think I found out why if failed above with `bash: $2: cannot assign in this way`.
"man bash" reads (excerpt)
```
Positional Parameters
... Positional parameters may not be assigned to with assignment statements.
```

Reply:
> because in general it is better to improve the functions instead of all its callers.

in general yes. but I think in case of quoting of parameters that can expand to empty strings or whitespace I would add quotes to the caller, because in general this can not be always compensated for inside the functions (imagine a function with more parameters, $scheme being the first of them - if it is empty, the others will shift - this is the case of `output_path()` and `backup_path()` functions here) and it is better to consistently give a correct example if the particular case does not strictly require it (other will copy and adapt the code). (I should have remembered this rule when writing the code and use quotes everywhere - I did not realize that `$scheme` may be empty.)

Reply:
> I think I found out why if failed above with `bash: $2: cannot assign in this way`.
> "man bash" reads (excerpt)
> 
> ```
> Positional Parameters
> ...
> Positional parameters may not be assigned to with assignment statements.
> ```

that must be it. Ans as in ReaR function arguments get quite consistently saved to local variables with more descriptive names at the start of every function (which is a good practice), it is not a big limitation.

Reply:
I fully agree with you that callers should call functions in a correct way.
I only meant that first and foremost the functions itself should be fail-safe.
Of couse I know that currently most functions in ReaR are not fail-safe.
E.g. a function that requires two arguments should check that it actually got
exactly two valid arguments so that when a caller provides something different
the function could do "the right thing".
What "the right thing" is depends on the particular function.

That bash parameter expansion with using a default value
is OK for our needs is what I meant with my comment
```
# but using a default value is practicable here because $1 is used only once
```
that is currently at
https://github.com/rear/rear/blob/jsmeix-fix-issue-2667/usr/share/rear/lib/global-functions.sh#L352